crypto/uri.c

```diff
@@ -0,0 +1,836 @@
+/*
+ * Copyright 2001-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <openssl/err.h>
+#include "internal/uri.h"
+#include "uri_charmap.h"
+
+#include <openssl/e_os2.h>
+
+/*
+ * pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+ *
+ * pct-encoded   = "%" HEXDIG HEXDIG
+ */
+static int span_pchars2(const char *p, size_t *off, const char *chars)
+{
+    size_t save_off = *off;
+
+    while (p[*off] != '\0') {
+        if (!is_valid(p[*off]))
+            break;
+        if (p[*off] == '%' && is_HEXDIG(p[*off + 1]) && is_HEXDIG(p[*off + 2]))
+            *off += 3;
+        else if (is_unreserved(p[*off]) || is_sub_delims(p[*off])
+                 || (chars != NULL && strchr(chars, p[*off]) != NULL))
+            (*off)++;
+        else
+            break;
+    }
+    return *off != save_off;
+}
+#define span_pchars(p, off) span_pchars2((p), (off), ":@")
+
+static int span_chars(const char *p, size_t *off, unsigned int types,
+                      const char *chars)
+{
+    size_t save_off = *off;
+
+    /*
+     * pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+     *
+     * pct-encoded   = "%" HEXDIG HEXDIG
+     */
+    while (p[*off] != '\0') {
+        if (!is_valid(p[*off]))
+            break;
+        if (is_type(p[*off], types)
+            || (chars != NULL && strchr(chars, p[*off]) != NULL))
+            (*off)++;
+        else
+            break;
+    }
+    return *off != save_off;
+}
+
+/*
+ * scheme        = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+ */
+static int extract_scheme(char **scheme, const char **pp)
+{
+    const char *p = *pp;
+    size_t off = 0;
+
+    *scheme = NULL;
+
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_SCHEME, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if (!is_ALPHA(p[off]))
+        return 1;
+    while (span_chars(p, &off, CHARTYPE_ALPHA|CHARTYPE_DIGIT, "+-."))
+        ;
```

Comment:
Okie.  I found all other similar ones and replaced added `continue` on them as well.